flexmeasures/data/models/planning/utils.py

```diff
@@ -110,16 +112,41 @@ def get_prices(
             first_event_start = price_df.first_valid_index()
             last_event_end = price_df.last_valid_index() + resolution
             current_app.logger.warning(
-                f"Prices partially unknown for planning window. Trimming planning window (from {query_window[0]} until {query_window[-1]}) to {first_event_start} until {last_event_end}."
+                f"Prices partially unknown for planning window (sensor {price_sensor.id}). Trimming planning window (from {query_window[0]} until {query_window[-1]}) to {first_event_start} until {last_event_end}."
             )
             query_window = (first_event_start, last_event_end)
         else:
             raise UnknownPricesException(
-                "Prices partially unknown for planning window."
+                f"Prices partially unknown for planning window (sensor {price_sensor.id})."
             )
     return price_df, query_window
 
 
+def inflexible_device_forecasts(
+    query_window: Tuple[datetime, datetime],
+    resolution: timedelta,
+    sensor: Sensor,
+):
+    """
+    :returns: power forecasts (consumption is positive, production is negative)
+    """
+    bdf: tb.BeliefsDataFrame = TimedBelief.search(
```

Comment:
Why are we not limiting the search to some horizon range which fits to forecasts?

Reply:
This is a great observation. But to me, a horizon range is not appropriate here. Instead, I'll implement a `belief_time` filter.

Right now the scheduler takes into account the most recent beliefs. So if the requested schedule lies in the future, the data here will consist of (the most recent) forecasts (if any exist). And if the requested schedule lies in the past, the data here will consist of (the most recent) measurements (if any exist). The latter amounts to answering "What if we could have scheduled under perfect foresight?".

We already extended our scheduling API with an optional `"prior"` field to specify a `belief_time` for recording the resulting schedule. Here, we should also use this field as a filter on what data should be used for the schedule, in order to support simulating scheduling over historical data.

Reply:
In the course of implementing this filter I found out that our test data for market prices was in UTC rather than in the Europe/Amsterdam timezone. I moved the data to the Amsterdam timezone, and adjusted other scheduling tests accordingly.

#### Explanation

The UTC prices had led to some unexpected test data points (especially their belief time, which is why I ran into it). We had set up hourly prices with a zero horizon, expecting each of the 24 prices to share a belief time of noon D-1 (as defined by our knowledge horizon function for the day-ahead market). But the belief time is determined using the Europe/Paris timezone instead of using the UTC timezone, so some of the prices suggested they had been recorded on a different day. Because of this, the scheduler unexpectedly trimmed the price window once I had set up the new `belief_time` filter.